### PR TITLE
Create generators flexibility

### LINKED ISSUE #206: create_generators flexibility
Currently we allow seeds with different scopes but it is not clear how a signature specific seed is going to be used since it is passed as a parameter to `create_generators` and not as input. Also i think this is a case that we want to reduce optionality (for example we don't want `generator_seed = hash(all_messages)`).

I think it's best to not allow "scoped seeds" but different `create_generators` operations. Each create_generators will accept a `seed` and a `dst` as a parameter and a `count` as input. If one wants signature specific generators they can use the `count` input but nothing else. This will be enough for "legacy support". 

This also makes it easier to describe the requirements of the `create_generators` operations (rather than the requirements for the seed selection).

IMO there are 2 options:
1. Don't allow different `create_generators` or `scoped_seeds`: No changes are needed, expect to remove the text around "scoped seeds".
2. Allow different `create_generators`: In this case IMO we should:
    - Add a `create_generators_operation` parameter in the ciphersuites definition.
    - For the suites defined in the document give the `create_generators_operation` parameter the value of "`create_generators` with `generator_seed=...`".
    - Require every create_generators operation to also define a unique ID which will be added to the ciphersuite_id (like EXP_TAG for expand_message in [h2c](https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-16.html#name-suite-id-naming-conventions)).
    - Add `create_generators_operation` as a parameter to Sign, Verify etc..
    - Add the requirements for defining a new `create_generators_operation`.

Obviously 1 is the simplest but the least flexible. It also is a "breaking change" with previous versions. So my preference will be 2 but I'm also very interested in others thoughts on the matter.

### PR DESCRIPTION
**Pull request by BasileiosKal**
Closes #206

Going with option 2 from the issue, allowing the definition of new `create_generators` and listing the relative requirements. Τhis will also benefit use cases like [Bound signatures](https://github.com/decentralized-identity/bbs-signature/issues/262).

### CODE REVIEW

**Comment by BasileiosKal** [comment]
Discussed on the WG call on the 29th of May, will re-consider the use of `extra_info` and add the PK as a parameter to the create_generators API.

**Review by Wind4Greg — APPROVED** [review]

**Comment by christianpaquin** [comment]
I'm not sure I understand the goal of parameterizing the generators creation. Isn't there a value for applications to pre-generate and reuse them across many issuers (which would not be possible of if they depend on an issuer PK).  Is the goal of the `header` to further scope the generators, so an issuer can have multiple distinct generators? Generators can be used to track users, so their uniformity has value.

**Comment by andrewwhitehead** [comment]
The ciphersuite does not have to take the PK and header into account when creating the generators, and the ones defined by the spec will not. It just seemed reasonable to pass those parameters in case the suite wants to define the generators differently (such as how the original implementation did, based on the issuer PK and number of messages). I don't think this increases correlation as the PK and header are known to all verifiers.

**Comment by andrewwhitehead** [comment]
It's the signature header, which is per issuance (and may be empty). It contributes to the domain hash.

**Comment by christianpaquin** [comment]
> It's the signature header, which is per issuance (and may be empty). It contributes to the domain hash.

Ok, so it doesn't add traceability risks then.

**Comment by BasileiosKal** [comment]
I think a malicious issuer could use a unique header per signature to track the user. There are plans for adding privacy considerations to address this. The user (or anyone else) could check the header and decide if it is conformant to those considerations. 

Indented use-cases for the header (e.g., to include algorithm/protocol identifiers etc.,) make it hard to define it as part of the ciphersuite.

**Comment by BasileiosKal** [comment]
Discussed on the WG call on the 19th of June. Will remove the header parameter from the Inputs of `create_generators`.

**Comment by christianpaquin** [comment]
> Discussed on the WG call on the 19th of June. Will remove the header parameter from the Inputs of `create_generators`.

The fixtures from PR #269 currently used the header, right?

File: draft-irtf-cfrg-bbs-signatures.md [hunk 1]
```diff
@@ -745,12 +766,29 @@ Procedure:
...
+- It should be infeasible to guess the discrete logarithm of the returned points, for any base, even with knowledge of the public parameters that were used to create those generators (like the `generator_seed` value in [Hash to Generators](#hash-to-generators)). Note that pseudo randomness does not necessarily imply this property. For example, an implementation that repeatably hashes a public seed value to create exponents `r_1, r_2, ..., r_count` (where `r_1 = hash(seed), r_2 = hash(r_1), ...`) and then returns the points `H_1 = P1 * r_1, H_2 = P_1 * r_2, ..., H_count = P_1 * r_count` would be insecure (given knowledge of the seed), but given knowledge of only the points `H_1, ..., H_count`, the sequence would appear random.
+- The returned points must be different from the Identity point of G1 as well as the constant point `P1` defined by the ciphersuite.
+- Must be constant time for a specific `count` value.
+- MUST be deterministic.
+- Must use proper domain separation for both the `create_generator` procedure, as well as all the internally called procedures.
```
Comment (andrewwhitehead) [inline draft-irtf-cfrg-bbs-signatures.md:785]:
```suggestion
- Must use proper domain separation for both the `create_generators` procedure, as well as all of the internally-called procedures.
```

**Review by andrewwhitehead — APPROVED** [review]

**Review by christianpaquin — APPROVED** [review]
LGTM

**Review by tplooker — APPROVED** [review]

**Comment by tplooker** [comment]
Discussed on WG call 26th of June, multiple approvals merging.

**Pull request merged**